make sure instances of c/leader_router are stopped nicely

bashtanov opened the pull request:
`cluster::leader_router` implements retry logic, which, depending on the parameters, may take quite a while to complete

Make sure it is always stopped quickly and correctly: its abort source is triggered early, and it is destroyed after all its operations complete.

Fixes slow shutdowns in https://redpandadata.atlassian.net/browse/CORE-7332

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [x] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bashtanov committed:
c/data-migrate/router: use controller abort source

`cluster::data_migration_router` is a wrapper on top of a couple of
instances of `cluster::leader_router`.

`cluster::leader_router` will search for partition leaders and direct
requests to appropriate nodes. If leader is not found it'll backoff and
retry, with current settings for up to 20 minutes altogether.

This may lead to slow shutdowns. Abort early.

bashtanov commented:
/ci-repeat 4
dt-repeat=25
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_creating_and_listing_migrations

bashtanov commented:
/ci-repeat 4
dt-repeat=25
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_creating_and_listing_migrations

bashtanov commented:
/ci-repeat 4
dt-repeat=25
tests/rptest/tests/data_migrations_api_test.py::DataMigrationsApiTest.test_creating_and_listing_migrations

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--39b98cbc-c92f-472f-b8b8-39934fa84200-->
<!--TEST RESULTS-->
<!--c954d53f-5de2-4043-b880-5a2bddf07af3-->
<!--TEST RESULTS-->
<!--9768dfa5-e9e3-45a6-835b-73bb3d820f15-->
<!--TEST RESULTS-->
<!--6530176f-1202-4275-8704-30f3b0b8e629-->
<!--TEST RESULTS-->
<!--a1c93ced-28fe-4d92-8a36-eecd06334d3d-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/76192">build#76192</a></summary>


| test_class            | test_method                          | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                             |
|:----------------------|:-------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest | test_creating_and_listing_migrations | null             | integration | https://buildkite.com/redpanda/redpanda/builds/76192#019a7c9b-629a-413b-b86c-cd07fe6990cf | FLAKY         | 44/45    | upstream reliability is '98.84105960264901'. current run reliability is '95.23809523809523'. drift is 3.60296 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations |
| DataMigrationsApiTest | test_creating_and_listing_migrations | null             | integration | https://buildkite.com/redpanda/redpanda/builds/76192#019a7ca0-8c67-4573-a66f-76849f5293ca | FLAKY         | 43/45    | upstream reliability is '99.6661101836394'. current run reliability is '90.9090909090909'. drift is 8.75702 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations |
| DataMigrationsApiTest | test_creating_and_listing_migrations | null             | integration | https://buildkite.com/redpanda/redpanda/builds/76192#019a7ca0-8c68-4030-b554-70be90afd32a | FLAKY         | 44/45    | upstream reliability is '98.84105960264901'. current run reliability is '95.23809523809523'. drift is 3.60296 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/76195">build#76195</a></summary>


| test_class                    | test_method                                  | test_arguments                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                             |
|:------------------------------|:---------------------------------------------|:----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests | test_replication_basic                       | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "redpanda"}} | integration | https://buildkite.com/redpanda/redpanda/builds/76195#019a7cd9-dae6-4c92-ae0a-de09d3f2159f | FLAKY         | 17/21    | upstream reliability is '88.14722395508421'. current run reliability is '80.95238095238095'. drift is 7.19484 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic       |
| ShadowLinkingReplicationTests | test_replication_basic                       | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "redpanda"}} | integration | https://buildkite.com/redpanda/redpanda/builds/76195#019a7cda-2e50-467a-a720-04fa913242a1 | FLAKY         | 15/21    | upstream reliability is '88.1619937694704'. current run reliability is '71.42857142857143'. drift is 16.73342 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic       |
| JavaCompressionTest           | test_upgrade_java_compression                | {"compression_type": "gzip"}                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/76195#019a7cd9-dae9-435e-a4e4-49235b6135ca | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression          |
| JavaCompressionTest           | test_upgrade_java_compression                | {"compression_type": "gzip"}                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/76195#019a7cda-2e52-4a13-81eb-1492f5a3b7b2 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression          |
| JavaCompressionTest           | test_upgrade_java_compression                | {"compression_type": "lz4"}                                                       | integration | https://buildkite.com/redpanda/redpanda/builds/76195#019a7cd9-daea-49bf-8ea1-17d223808296 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression          |
| JavaCompressionTest           | test_upgrade_java_compression                | {"compression_type": "lz4"}                                                       | integration | https://buildkite.com/redpanda/redpanda/builds/76195#019a7cda-2e53-4277-8e70-44ee18cc0478 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression          |
| JavaCompressionTest           | test_upgrade_java_compression                | {"compression_type": "snappy"}                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/76195#019a7cd9-daec-40e7-b073-dff6e7f76729 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression          |
| JavaCompressionTest           | test_upgrade_java_compression                | {"compression_type": "snappy"}                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/76195#019a7cda-2e54-421b-95a5-057378d70d79 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression          |
| JavaCompressionTest           | test_upgrade_java_compression                | {"compression_type": "zstd"}                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/76195#019a7cd9-daed-4f99-867b-5aa266557b5b | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression          |
| JavaCompressionTest           | test_upgrade_java_compression                | {"compression_type": "zstd"}                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/76195#019a7cda-2e55-4de7-a1f1-3bc9d3c1846e | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=JavaCompressionTest&test_method=test_upgrade_java_compression          |
| MountUnmountIcebergTest       | test_simple_remount                          | {"cloud_storage_type": 1}                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/76195#019a7cd9-dae2-40e7-9eec-f49d3e390c8d | FLAKY         | 19/21    | upstream reliability is '85.21008403361344'. current run reliability is '90.47619047619048'. drift is -5.26611 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount                |
| SegmentMsTest                 | test_segment_rolling_with_retention_consumer | null                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/76195#019a7cd9-dae3-43da-9631-558fd2379c67 | FLAKY         | 19/21    | upstream reliability is '92.75053304904051'. current run reliability is '90.47619047619048'. drift is 2.27434 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer |
| SegmentMsTest                 | test_segment_rolling_with_retention_consumer | null                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/76195#019a7cda-2e4d-4b42-8026-699b08f5d1ca | FLAKY         | 19/21    | upstream reliability is '92.75053304904051'. current run reliability is '90.47619047619048'. drift is 2.27434 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/76389">build#76389</a></summary>


| test_class              | test_method                                  | test_arguments                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                               |
|:------------------------|:---------------------------------------------|:---------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| AuditLogTestKafkaApi    | test_no_auth_enabled                         | {"audit_transport_mode": "kclient"}                                                                            | integration | https://buildkite.com/redpanda/redpanda/builds/76389#019a84a7-d78f-447d-97e2-c44ae202fcfd | FLAKY         | 17/21    | upstream reliability is '96.66666666666667'. current run reliability is '80.95238095238095'. drift is 15.71429 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AuditLogTestKafkaApi&test_method=test_no_auth_enabled                    |
| EndToEndCloudTopicsTest | test_delete_records                          | null                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/76389#019a84a7-d788-401b-9d30-08f734b5734a | FAIL          | 0/1      |                                                                                                                                                                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndCloudTopicsTest&test_method=test_delete_records                  |
| DataMigrationsApiTest   | test_conflicting_names                       | null                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/76389#019a849d-e574-4895-8842-44d581c3eff9 | FAIL          | 0/1      |                                                                                                                                                                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_conflicting_names                 |
| DataMigrationsApiTest   | test_listing_inexistent_migration            | null                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/76389#019a849d-e574-4895-8842-44d581c3eff9 | FAIL          | 0/1      |                                                                                                                                                                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_listing_inexistent_migration      |
| MountUnmountIcebergTest | test_simple_remount                          | {"cloud_storage_type": 1}                                                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/76389#019a84a7-d789-4dd0-b861-4998235dfb86 | FLAKY         | 17/21    | upstream reliability is '80.69087688219663'. current run reliability is '80.95238095238095'. drift is -0.2615 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount                  |
| SchemaScaleTest         | schema_scale_test                            | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "spark", "use_partition_spec": false} | integration | https://buildkite.com/redpanda/redpanda/builds/76389#019a849d-e578-40ee-b8cf-b4f37aae9ecc | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SchemaScaleTest&test_method=schema_scale_test                            |
| DeleteRecordsTest       | test_delete_records_concurrent_truncations   | {"cloud_storage_enabled": true, "truncate_point": "at_high_watermark"}                                         | integration | https://buildkite.com/redpanda/redpanda/builds/76389#019a84a7-d788-401b-9d30-08f734b5734a | FAIL          | 0/1      |                                                                                                                                                                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DeleteRecordsTest&test_method=test_delete_records_concurrent_truncations |
| SegmentMsTest           | test_segment_rolling_with_retention_consumer | null                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/76389#019a849d-e578-40ee-b8cf-b4f37aae9ecc | FLAKY         | 19/21    | upstream reliability is '92.63848396501457'. current run reliability is '90.47619047619048'. drift is 2.16229 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer   |
| NodeWiseRecoveryTest    | test_node_wise_recovery                      | {"dead_node_count": 2}                                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/76389#019a849d-e574-4895-8842-44d581c3eff9 | FAIL          | 0/1      |                                                                                                                                                                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodeWiseRecoveryTest&test_method=test_node_wise_recovery                 |
| SchemaRegistryTest      | test_serde_client                            | {"client_type": 3, "protocol": 1}                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/76389#019a849d-e574-4895-8842-44d581c3eff9 | FAIL          | 0/1      |                                                                                                                                                                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SchemaRegistryTest&test_method=test_serde_client                         |
| SelfTestTest            | test_self_test                               | null                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/76389#019a849d-e578-40ee-b8cf-b4f37aae9ecc | FLAKY         | 16/21    | upstream reliability is '100.0'. current run reliability is '76.19047619047619'. drift is 23.80952 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SelfTestTest&test_method=test_self_test                                  |
| CreateTopicsTest        | test_case_insensitive_boolean_property       | null                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/76389#019a849d-e578-40ee-b8cf-b4f37aae9ecc | FLAKY         | 16/21    | upstream reliability is '100.0'. current run reliability is '76.19047619047619'. drift is 23.80952 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=CreateTopicsTest&test_method=test_case_insensitive_boolean_property      |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/76435">build#76435</a></summary>


| test_class                              | test_method                        | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                             |
|:----------------------------------------|:-----------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ClusterLinkingTopicSyncingWithTlsValues | test_topic_syncing_include_exclude | null             | integration | https://buildkite.com/redpanda/redpanda/builds/76435#019a8d5c-0103-442d-a10b-013323462b7d | FLAKY         | 18/21    | upstream reliability is '100.0'. current run reliability is '85.71428571428571'. drift is 14.28571 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterLinkingTopicSyncingWithTlsValues&test_method=test_topic_syncing_include_exclude |
| LogCompactionTxRemovalTest              | test_tx_control_batch_removal      | null             | integration | https://buildkite.com/redpanda/redpanda/builds/76435#019a8d5c-0103-442d-a10b-013323462b7d | FLAKY         | 18/21    | upstream reliability is '83.33333333333334'. current run reliability is '85.71428571428571'. drift is -2.38095 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTxRemovalTest&test_method=test_tx_control_batch_removal                   |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/76892">build#76892</a></summary>


| test_class                     | test_method                                | test_arguments                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                                            |
|:-------------------------------|:-------------------------------------------|:--------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ClusterConfigLegacyDefaultTest | test_legacy_default_explicit_after_upgrade | {"wipe_cache": true}                                                            | integration | https://buildkite.com/redpanda/redpanda/builds/76892#019ab6f1-84a0-461f-af14-6c9536b3423a | FAIL          | 0/1      |                                                                                                                                                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterConfigLegacyDefaultTest&test_method=test_legacy_default_explicit_after_upgrade |
| ShadowLinkingReplicationTests  | test_replication_with_failures             | null                                                                            | integration | https://buildkite.com/redpanda/redpanda/builds/76892#019ab6e7-7c25-45b9-96cb-d1625f4a4076 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_with_failures              |
| DataMigrationsApiTest          | test_conflicting_names                     | null                                                                            | integration | https://buildkite.com/redpanda/redpanda/builds/76892#019ab6f1-84a0-461f-af14-6c9536b3423a | FAIL          | 0/1      |                                                                                                                                                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_conflicting_names                              |
| DatalakeE2ETests               | test_avro_schema                           | {"catalog_type": "nessie", "cloud_storage_type": 1, "query_engine": "spark"}    | integration | https://buildkite.com/redpanda/redpanda/builds/76892#019ab6f1-84a0-461f-af14-6c9536b3423a | FAIL          | 0/1      |                                                                                                                                                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeE2ETests&test_method=test_avro_schema                                         |
| DatalakeE2ETests               | test_json_schema                           | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1, "query_engine": "spark"} | integration | https://buildkite.com/redpanda/redpanda/builds/76892#019ab6f1-84a0-461f-af14-6c9536b3423a | FAIL          | 0/1      |                                                                                                                                                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeE2ETests&test_method=test_json_schema                                         |
| SchemaEvolutionE2ETests        | test_schema_evolution                      | {"cloud_storage_type": 1, "qe_and_cat": ["spark", "rest_jdbc"]}                 | integration | https://buildkite.com/redpanda/redpanda/builds/76892#019ab6f1-84a0-461f-af14-6c9536b3423a | FAIL          | 0/1      |                                                                                                                                                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SchemaEvolutionE2ETests&test_method=test_schema_evolution                             |
| ShardPlacementTest             | test_core_count_change                     | null                                                                            | integration | https://buildkite.com/redpanda/redpanda/builds/76892#019ab6f1-84a5-4d6d-9a93-a924bba12d11 | FLAKY         | 20/21    | upstream reliability is '99.75062344139651'. current run reliability is '95.23809523809523'. drift is 4.51253 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShardPlacementTest&test_method=test_core_count_change                                 |
</details>
<!--/body-->

bashtanov commented:
/ci-repeat 1

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/76195-->
<!--0859727c-fa4e-40f5-b8c7-9b90a80f1af1-->
<!--c9384cce-46ae-4c37-b8ba-3a8510f4fd75-->
<!--16ff5e8d-bcfd-4be6-826a-9b6d959ad4e2-->
<!--8a3ab165-f441-4e36-92e9-7e0554bde38c-->
<!--d4cb5088-f9a1-4219-ba3b-53762b0ed2c2-->
<!--60317c63-ec49-4189-9893-2ce8761e8bcf-->
<!--51a6d950-0e31-4525-85fb-35aca84159ce-->
<!--7cb72499-4358-4462-b06c-4a79a071a4a4-->
<!--title-->
## Retry command for [Build#76195](https://buildkite.com/redpanda/redpanda/builds/76195)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/compatibility/java_compression_test.py::JavaCompressionTest.test_upgrade_java_compression@{"compression_type":"zstd"}
tests/rptest/tests/compatibility/java_compression_test.py::JavaCompressionTest.test_upgrade_java_compression@{"compression_type":"lz4"}
tests/rptest/tests/compatibility/java_compression_test.py::JavaCompressionTest.test_upgrade_java_compression@{"compression_type":"snappy"}
tests/rptest/tests/compatibility/java_compression_test.py::JavaCompressionTest.test_upgrade_java_compression@{"compression_type":"gzip"}
```
<!--/body-->

bashtanov commented:
failures are unrelated, please force-merge if approving

joe-redpanda commented on src/v/cluster/data_migration_router.cc:
```diff
@@ -61,7 +63,11 @@ router::router(
-      max_backoff) {}
+      max_backoff)
+  , _as_sub(ssx::subscribe_or_trigger(as, [this] noexcept {
+      _get_group_offsets_router.request_stop();
```
Should we be awaiting stop of these?
edit:
in a stop method on data_migration_router

bashtanov replied:
Yes, you are right

joe-redpanda replied:
I'll drop the commit with stop in mine if you want to add it here

bashtanov replied:
I will add it here. It also uncovers a few more occasions where we don't stop leader_router properly!

bashtanov committed:
cluster: stop leader_routers for them to wait their gates

bashtanov commented on src/v/cluster/data_migration_router.cc:
```diff
@@ -61,7 +63,16 @@ router::router(
-      max_backoff) {}
+      max_backoff)
+  , _as_sub(ssx::subscribe_or_trigger(as, [this] noexcept {
+      _get_group_offsets_router.request_stop();
+      _set_group_offsets_router.request_stop();
```
it does not matter

bashtanov commented:
@andrwng @mmaslankaprv do you think the non-data-migrations stuff is worth backporting?

bashtanov committed:
c/migrations/tx_manager_migrator{,_handler}: use app abort source

A component serving admin api requests should be aborted before admin
api is drained.
Similary, a component serving RPC requests should be aborted before RPC
request is drained.

Otherwise shutdown may get stuck if there is an RPC or an Admin API
request in flight.

bashtanov committed:
c/migrations/tx_manager_migrator: use ungated c/leader_router

RPC handlers are not designed to have an async stop(), but one of them,
`tx_manager_migrator_handler`, uses leader routers. It is safe, as RPC
server is drained before the handlers are destroyed.

Introduce an ungated flavour of cluster::leader_router for such use.
It's also going to be used in tx_manager_migrator itself, as it uses
the same leader_router subclasses. It's also safe, as its use is led by
Admin API, which is drained before tx_manager_migrator.

bashtanov commented:
/dt

mmaslankaprv commented on src/v/cluster/leader_router.h:
```diff
@@ -58,6 +61,12 @@ concept IsHandler = requires(req_t req, resp_t resp, handler_t handler) {
 
+// For to use a leader_router in a component without async stop(). Such
+// components must make sure to drain the router before destructing it.
+struct leader_router_ungated_tag;
```
do we really need this ? It seems weird to rely on a fragile order of components shutdown. I feels like an antipattern to me that in order to make it stop we must have an ungated version

bashtanov replied:
I didn't like it either, but the alternatives are the following:
a) Have the gate but don't close it (like before) -- even worse, hides the problem.
b) Implement asynchronous stop for rpc::rpc_server and its handlers. Might be a decent thing to do, but bloats the scope of the PR further more.

mmaslankaprv replied:
I thought the RPC servers are already stopped in an asynchonous way

bashtanov replied:
Well, servers are, but not handlers. `rpc::rpc_server` does not have a `stop()` method that would propagate to entries of `_services`.

Instead, it just destructs them. It does it only after calling `net::server::wait_for_shutdown` though. It waits for `_conn_gate`, which is held when handled. So RPC server guarantees that RPC handler methods it calls never outlive the handler object.
